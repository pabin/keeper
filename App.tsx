import React, { useState, useEffect } from 'react';
import { useColorScheme } from 'react-native';
import Toast from 'react-native-toast-message';

import { getObjectData } from './src/model/storageUtils';
import {
  createOrUpdateNoteDetails,
  updateNoteStatus,
} from './src/model/noteModel';
import StackNavigator from './src/navigation/StackNavigation';
import { themes } from './src/styles/colors';
import { Note } from './src/types/note';
import { NoteContextInterface } from './src/types/noteContext';

export const NoteContext = React.createContext<NoteContextInterface>([]);

const App = (): JSX.Element => {
  const [loading, setLoading] = useState<boolean>(true);
  const [notes, setNotes] = useState<Note[]>([]);
  const [favouriteNotes, setFavouriteNotes] = useState<Note[]>([]);
  const [archivedNotes, setArchivedNotes] = useState<Note[]>([]);

  // To Do: Theme switch
  const isDarkMode = useColorScheme() === 'dark';

  const getDatabase = async () => {
    const notesData: Note[] = await getObjectData('notes');

    if (notesData) {
      const favourites = notesData.filter(n => n.isFavourite);
      const archived = notesData.filter(n => n.isArchived);
      setNotes(notesData);
      setArchivedNotes(archived);
      setFavouriteNotes(favourites);
      setLoading(false);
    }
  };

  useEffect(() => {
    getDatabase();
  }, []);

  const onNoteStatusUpdate = async (note: Note, action: string) => {
    await updateNoteStatus(note, action);
    await getDatabase();
  };

  const onCreateOrUpdateNote = async (
    note: Note,
    title: string,
    body: string,
  ) => {
    await createOrUpdateNoteDetails(note, title, body);
    await getDatabase();
  };

  return (
    <>
      <NoteContext.Provider
        value={{
          notes,
          loading,
          favouriteNotes,
          archivedNotes,
          onNoteStatusUpdate,
          onCreateOrUpdateNote,
          theme: isDarkMode ? themes.dark : themes.light,
        }}
      >
        <StackNavigator />
        <Toast />
      </NoteContext.Provider>
    </>
  );
};

export default App;
