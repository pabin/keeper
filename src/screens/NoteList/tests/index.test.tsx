import React from 'react';
import { cleanup, render } from '@testing-library/react-native';

import NoteListScreen from '../index';
import { notes } from '../../../fixtures/notes';
import { NoteContext } from '../../../../App';

describe('NoteListScreen', () => {
  afterEach(cleanup);
  const navigateMock = jest.fn();

  const navigation = {
    navigate: navigateMock,
  };
  const renderComponent = (notes, loading) => {
    return render(
      <NoteContext.Provider value={{ notes, loading }}>
        <NoteListScreen navigation={navigation} />
      </NoteContext.Provider>,
    );
  };

  test('renders correctly when notes are loading', () => {
    const { toJSON } = renderComponent([], true);

    expect(toJSON()).toMatchSnapshot();
  });

  test('renders correctly when notes are loaded', () => {
    const { toJSON } = renderComponent(notes, false);

    expect(toJSON()).toMatchSnapshot();
  });

  test('renders correctly when no notes are present', () => {
    const { toJSON } = renderComponent([], false);

    expect(toJSON()).toMatchSnapshot();
  });
});
