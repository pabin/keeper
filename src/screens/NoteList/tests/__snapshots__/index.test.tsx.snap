// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteListScreen renders correctly when no notes are present 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#01579B",
        "borderRadius": 30,
        "bottom": 24,
        "height": 60,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 24,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.4,
        "width": 60,
        "zIndex": 3,
      }
    }
    testID="createNoteIcon"
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 24,
          },
          undefined,
          Object {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#455A64",
            "fontSize": 32,
          },
          undefined,
          Object {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
    <Text
      style={
        Object {
          "color": "#455A64",
          "fontSize": 18,
          "marginVertical": 12,
        }
      }
    >
      You don't have any notes!
    </Text>
  </View>
</View>
`;

exports[`NoteListScreen renders correctly when notes are loaded 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#01579B",
        "borderRadius": 30,
        "bottom": 24,
        "height": 60,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 24,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.4,
        "width": 60,
        "zIndex": 3,
      }
    }
    testID="createNoteIcon"
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 24,
          },
          undefined,
          Object {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "body": "**this is test note**
\`this is new lne\`",
          "createdAt": 2022-01-21T05:49:45.613Z,
          "id": 100,
          "isArchived": false,
          "isFavourite": false,
          "title": "note title",
          "updatedAt": 2022-01-21T05:52:58.634Z,
        },
        Object {
          "body": "**this is test note 2**
\`this is new lne\`",
          "createdAt": 2022-01-21T05:49:45.613Z,
          "id": 101,
          "isArchived": false,
          "isFavourite": false,
          "title": "note title 2",
          "updatedAt": 2022-01-21T05:52:58.634Z,
        },
        Object {
          "body": "**this is test note 3**
\`this is new lne\`",
          "createdAt": 2022-01-21T05:49:45.613Z,
          "id": 102,
          "isArchived": false,
          "isFavourite": false,
          "title": "note title 3",
          "updatedAt": 2022-01-21T05:52:58.634Z,
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEnabled={true}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "borderRadius": 12,
              "flex": 1,
              "flexDirection": "row",
              "margin": 12,
              "marginVertical": 8,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 6,
                "width": 0,
              },
              "shadowOpacity": 0.2,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#D6DBDF",
                "borderBottomLeftRadius": 12,
                "borderTopLeftRadius": 12,
                "flex": 4,
                "flexDirection": "row",
                "opacity": 1,
                "padding": 16,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 6,
                  "width": 0,
                },
                "shadowOpacity": 0.1,
              }
            }
            testID="detailNavigator"
          >
            <View
              style={
                Object {
                  "marginRight": 12,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#607D8B",
                      "fontSize": 24,
                    },
                    undefined,
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
            <View>
              <Text
                style={
                  Object {
                    "color": "#222",
                    "fontSize": 16,
                    "fontWeight": "600",
                  }
                }
              >
                note title
              </Text>
              <Text
                style={
                  Object {
                    "color": "#222",
                    "fontSize": 12,
                  }
                }
              >
                Fri Jan 21 2022 | 11:34:45 AM
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "opacity": 1,
              }
            }
            testID="createNoteNavigator"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#01579B",
                    "fontSize": 22,
                  },
                  undefined,
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "borderRadius": 12,
              "flex": 1,
              "flexDirection": "row",
              "margin": 12,
              "marginVertical": 8,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 6,
                "width": 0,
              },
              "shadowOpacity": 0.2,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#D6DBDF",
                "borderBottomLeftRadius": 12,
                "borderTopLeftRadius": 12,
                "flex": 4,
                "flexDirection": "row",
                "opacity": 1,
                "padding": 16,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 6,
                  "width": 0,
                },
                "shadowOpacity": 0.1,
              }
            }
            testID="detailNavigator"
          >
            <View
              style={
                Object {
                  "marginRight": 12,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#607D8B",
                      "fontSize": 24,
                    },
                    undefined,
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
            <View>
              <Text
                style={
                  Object {
                    "color": "#222",
                    "fontSize": 16,
                    "fontWeight": "600",
                  }
                }
              >
                note title 2
              </Text>
              <Text
                style={
                  Object {
                    "color": "#222",
                    "fontSize": 12,
                  }
                }
              >
                Fri Jan 21 2022 | 11:34:45 AM
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "opacity": 1,
              }
            }
            testID="createNoteNavigator"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#01579B",
                    "fontSize": 22,
                  },
                  undefined,
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF",
              "borderRadius": 12,
              "flex": 1,
              "flexDirection": "row",
              "margin": 12,
              "marginVertical": 8,
              "shadowColor": "#000",
              "shadowOffset": Object {
                "height": 6,
                "width": 0,
              },
              "shadowOpacity": 0.2,
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#D6DBDF",
                "borderBottomLeftRadius": 12,
                "borderTopLeftRadius": 12,
                "flex": 4,
                "flexDirection": "row",
                "opacity": 1,
                "padding": 16,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 6,
                  "width": 0,
                },
                "shadowOpacity": 0.1,
              }
            }
            testID="detailNavigator"
          >
            <View
              style={
                Object {
                  "marginRight": 12,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#607D8B",
                      "fontSize": 24,
                    },
                    undefined,
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
            <View>
              <Text
                style={
                  Object {
                    "color": "#222",
                    "fontSize": 16,
                    "fontWeight": "600",
                  }
                }
              >
                note title 3
              </Text>
              <Text
                style={
                  Object {
                    "color": "#222",
                    "fontSize": 12,
                  }
                }
              >
                Fri Jan 21 2022 | 11:34:45 AM
              </Text>
            </View>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "opacity": 1,
              }
            }
            testID="createNoteNavigator"
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#01579B",
                    "fontSize": 22,
                  },
                  undefined,
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`NoteListScreen renders correctly when notes are loading 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flex": 1,
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#01579B",
        "borderRadius": 30,
        "bottom": 24,
        "height": 60,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 24,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 4,
          "width": 0,
        },
        "shadowOpacity": 0.4,
        "width": 60,
        "zIndex": 3,
      }
    }
    testID="createNoteIcon"
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 24,
          },
          undefined,
          Object {
            "fontFamily": "FontAwesome",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <ActivityIndicator
      color="#01579B"
      size="small"
    />
  </View>
</View>
`;
